File: 백준/Gold/9663. N－Queen/N－Queen.cc
#include <iostream>
#include <cmath>
#include <vector>

using namespace std;

int n, answer = 0;
int board[20][20] = {0,};
vector<pair<int, int>> queen; // 퀸의 위치 저장

bool attack(int x, int y) {
    for (int s = 0; s < queen.size(); s++) {
        int i = queen[s].first;
        int j = queen[s].second;
        
        if (x == i) // 퀸 수평 상에 존재
            return true;
        if (y == j) // 퀸 수직 상에 존재
            return true;

        if (abs(x - i) == abs(y - j)) // 퀸 대각선 상에 존재
            return true;
    }
    
    return false;
}

void search(int x, int y) {
    if (x == n - 1) {
        answer += 1;
        return;
    }
    
    // 다음 행에 퀸을 놓을 자리 체크
    int i = x + 1;
    for (int j = 0; j < n; j++) {
        if (board[i][j] != 0)
            continue;
        if (attack(i, j))
            continue;
        
        board[i][j] = 8;
        queen.push_back(make_pair(i, j));
        search(i, j);
        board[i][j] = 0;
        queen.pop_back();
    }
}

int main(void) {
    cin.tie(NULL); ios_base::sync_with_stdio(false);
    
    cin >> n;

    // 퀸은 수평, 수직, 대각선 이동 가능
    for (int i = 0; i < n; i++) {
        fill(&board[0][0], &board[19][20], 0); // 0: 놓을 수 있음, 1: 놓을 수 없음
        queen.push_back(make_pair(0, i));
        board[0][i] = 8;
        search(0, i);
        queen.pop_back();
        board[0][i] = 0;
    }
    cout << answer << "\n";

    return 0;
}

File: 백준/Gold/9663. N－Queen/N－Queen.py
n = int(input())
answer = 0
row = [0 for _ in range(n)]

def attack(x):
    for i in range(x):
        if row[x] == row[i]: # 수직 체크
            return True
        if abs(row[x] - row[i]) == abs(x - i): # 대각선 체크
            return True
    return False

def search(x):
    if x == n:
        global answer
        answer += 1
        return
    
    # 다음 행 체크
    for i in range(n):
        # [x, i]에 퀸
        row[x] = i
        if attack(x) == False: # 공격 가능한 위치 x
            search(x + 1)
        
search(0)

print(answer)